src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -325,6 +356,459 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+		}
+
+		(*commonEqClass)->equivalentVars = lappend((*commonEqClass)->equivalentVars,
+												   newEqMember);
+	}
+}
+
+
+/*
+ * TODO: (Andres asked) could the rte_relation that we're adding the eq.
+ * be part of ANTI-JOIN?
+ *
+ * AddToVarEquivalenceClass add the varToBeAdded to the given varEquivalanceClass.
+ * The function add the varToBeAdd if it belongs to a relation. If the varToBeAdded
+ * belongs to a subquery, the function recursively calls itself until it finds the
+ * relation that the var actually belongs to.
+ */
+static void
+AddToVarEquivalenceClass(PlannerInfo *root, Var *varToBeAdded,
+						 VarEquivalenceClass **varEquivalanceClass)
+{
+	RangeTblEntry *rte = root->simple_rte_array[varToBeAdded->varno];
+
+	if (rte->rtekind == RTE_RELATION)
+	{
+		VarEquivalenceClassMember *varEqMember = palloc0(
+			sizeof(VarEquivalenceClassMember));
+		varEqMember->varattno = varToBeAdded->varattno;
+		varEqMember->varno = varToBeAdded->varno;
+		varEqMember->relationId = rte->relid;
+
+		(*varEquivalanceClass)->equivalentVars =
+			lappend((*varEquivalanceClass)->equivalentVars, varEqMember);
+	}
+	else if (rte->rtekind == RTE_SUBQUERY && !rte->inh)
+	{
+		Query *subquery = rte->subquery;
+		RelOptInfo *baseRelOptInfo = NULL;
+		TargetEntry *subqueryTargetEntry = NULL;
+
+		/* punt if it's a whole-row var rather than a plain column reference */
+		if (varToBeAdded->varattno == InvalidAttrNumber)
+		{
+			return;
+		}
+
+		baseRelOptInfo = find_base_rel(root, varToBeAdded->varno);
+
+		/* If the subquery hasn't been planned yet, we have to punt */
+		if (baseRelOptInfo->subroot == NULL)
+		{
+			return;
+		}
+
+		Assert(IsA(baseRelOptInfo->subroot, PlannerInfo));
+
+		subquery = baseRelOptInfo->subroot->parse;
+		Assert(IsA(subquery, Query));
+
+		/* Get the subquery output expression referenced by the upper Var */
+		subqueryTargetEntry = get_tle_by_resno(subquery->targetList,
+											   varToBeAdded->varattno);
+		if (subqueryTargetEntry == NULL || subqueryTargetEntry->resjunk)
+		{
+			elog(ERROR, "subquery %s does not have attribute %d",
+				 rte->eref->aliasname, varToBeAdded->varattno);
+		}
+
+		if (!IsA(subqueryTargetEntry->expr, Var))
+		{
+			return;
+		}
+
+		varToBeAdded = (Var *) subqueryTargetEntry->expr;
+
+		/* Can only handle a simple Var of subquery's query level */
+		if (varToBeAdded && IsA(varToBeAdded, Var) && varToBeAdded->varlevelsup == 0)
+		{
+			AddToVarEquivalenceClass(baseRelOptInfo->subroot, varToBeAdded,
+									 varEquivalanceClass);
+		}
+	}
+}
+
+
+/*
+ * VarClassMemberEqualsToVarClass returns true if it has any equivalent member
+ * in the given varEqClass.
+ */
+static bool
+VarClassMemberEqualsToVarClass(VarEquivalenceClassMember *inputMember,
+							   VarEquivalenceClass *varEqClass)
+{
+	ListCell *classCell = NULL;
+	foreach(classCell, varEqClass->equivalentVars)
+	{
+		VarEquivalenceClassMember *memberOfClass = lfirst(classCell);
+		if (memberOfClass->varno == inputMember->varno &&
```

Comment:
comparing varno's could be problematic here. Same relationid used in 2 different subqueries and at the same location in rtable list would have the same varno. Does it make a difference in the correctness of this function ?